tempodb/blocklist/poller.go

```diff
@@ -151,44 +154,90 @@ func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
 		return nil, nil, err
 	}
 
-	blocklist := PerTenant{}
-	compactedBlocklist := PerTenantCompacted{}
+	var (
+		wg  = boundedwaitgroup.New(p.cfg.TenantPollConcurrency)
+		mtx = sync.Mutex{}
+
+		blocklist          = PerTenant{}
+		compactedBlocklist = PerTenantCompacted{}
 
-	consecutiveErrors := 0
+		finalErr       atomic.Error
+		tenantFailures atomic.Int32
+
+		tenantFailuresRemaining = p.cfg.TolerateTenantFailures
+	)
 
 	for _, tenantID := range tenants {
-		newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
-		if err != nil {
-			level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
-			consecutiveErrors++
-			if consecutiveErrors > p.cfg.TolerateConsecutiveErrors {
-				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
-				return nil, nil, err
+		// Exit early if we have exceeded our tolerance for number of failing tenants.
+		if tenantFailures.Load() >= int32(tenantFailuresRemaining) {
```

Comment:
i don't think you need tenantFailuresRemaining. i believe the check on line 237 is correct. (comparing number of failed tenants vs the config'ed value).

there's also a race on tenantFailuresRemaining b/c you're checking it here not under lock and modifying it  on line 208.

final question:
previously we'd attempt to poll a tenant multiple times based on the consecutive errors config. if any tenant failed the configurable number of times we'd return an error from this func. this change will make it so a tenant can fail to poll and this function doesn't return an error. 

i'm fine with this behavior change if we feel like it is the right choice operationally.

Reply:
The check at the top is to break early, so we don't continue to try and poll tenants after we exceeded our tolerance for failing tenants.  The check at the bottom is so that we catch exceeding the threshold if it happens on the last tenant.  I think we need both, but let's chat if you're seeing something else.  Maybe the one at the end is questionable since if we failed on the last tenant, perhaps better behavior is to return what we have worked for?

Nice, I see the race, I'll add another atomic for this.

One slight correction; previously there was no retry on a single tenant.  A tenant would fail, and we would store the previous results.  The previous early exit condition was based on the number of tenants failing in a row.  If more than one tenant failed (default tolerance of 1 failure) in a row then we would return an error.  Now we both retry failed tenants once by default, and we tolerate one failed tenant.  Operationally I like this pretty well, since if a tenant fails and then the retry succeeds, it doesn't stop the show.  Each tenant could fail once, but as long as they all succeed on the retry, we still have results.

What do you think about the change in behavior?  We can chat this afternoon.